glean-core/src/metrics/timing_distribution.rs

```diff
@@ -576,6 +577,141 @@ impl TimingDistributionMetric {
             test_get_num_recorded_errors(glean, self.meta(), error).unwrap_or(0)
         })
     }
+
+    /// Start a new histogram buffer associated with this timing distribution metric.
+    ///
+    /// A histogram buffer accumulates in-memory.
+    /// Data is recorded into the metric on drop.
+    pub fn start_buffer(&self) -> HistogramBuffer<'_> {
+        HistogramBuffer::new(self)
+    }
+
+    /// Start a new histogram buffer.
+    ///
+    /// A histogram buffer accumulates in-memory.
+    /// Data needs to be recorded into the metric using
+    /// [`TimingDistributionMetric::ffi_commit_buffer`].
+    #[doc(hidden)]
+    pub fn ffi_start_buffer(&self) -> Arc<FfiHistogramBuffer> {
+        Arc::new(FfiHistogramBuffer::new())
+    }
+
+    /// Record data from the histogram buffer into this metric.
+    pub fn ffi_commit_buffer(&self, buffer: Arc<FfiHistogramBuffer>) {
+        let metric = self.clone();
+        crate::launch_with_glean(move |glean| {
+            glean
+                .storage()
+                .record_with(glean, &metric.meta, move |old_value| {
+                    let mut hist = match old_value {
+                        Some(Metric::TimingDistribution(hist)) => hist,
+                        _ => Histogram::functional(LOG_BASE, BUCKETS_PER_MAGNITUDE),
+                    };
+
+                    let histogram = buffer.histogram.lock().unwrap();
+                    hist.merge(&histogram);
+                    Metric::TimingDistribution(hist)
+                });
+        });
+    }
+
+    fn commit_histogram(&self, histogram: Histogram<Functional>) {
+        let metric = self.clone();
+        crate::launch_with_glean(move |glean| {
+            glean
+                .storage()
+                .record_with(glean, &metric.meta, move |old_value| {
+                    let mut hist = match old_value {
+                        Some(Metric::TimingDistribution(hist)) => hist,
+                        _ => Histogram::functional(LOG_BASE, BUCKETS_PER_MAGNITUDE),
+                    };
+
+                    hist.merge(&histogram);
```

Comment:
All this moving of `histogram` across closures just so we can take a reference.

Reply:
We need to hold onto the data somewhere. But it doesn't make much sense for `merge` to consume it in the end, because we cannot merge two independent allocations anyway.